### DISCUSSION TITLE
Are there any API end points for WebAuthn in Keycloak that I can use to consume in a native application?

### DISCUSSION THREAD

**Discussion by Mohitkiran** [keycloak/keycloak#22590] [Q&A]
Hi People,

I am trying to develop Webauthn as 2nd factor after password for a native windows WPF application and I am trying to utilise Keycloaks WebAuthn capabilities to achieve this.

Are there any WebAuthn API's that I can use to accomplish this? Honestly I tried going through the documentation but didn't find any. Can some one please point me in right direction?

**Comment by justin-tay** [comment]
The client portion of WebAuthn is implemented in a browser. For Keycloak you would just need to configure the authentication and create a custom browser webauth flow that you bind to the Browser flow. There is a tutorial on how to setup such a flow at https://keycloak.ch/keycloak-tutorials/tutorial-webauthn/.

The best practice for implementing a native client is detailed in [RFC 8252 OAuth 2.0 for Native Apps](https://datatracker.ietf.org/doc/html/rfc8252) and there is a section on [Windows Implementation Details](https://datatracker.ietf.org/doc/html/rfc8252#appendix-B.3).

```
   OAuth 2.0 authorization requests from native apps should only be made
   through external user-agents, primarily the user's browser.  This
   specification details the security and usability reasons why this is
   the case and how native apps and authorization servers can implement
   this best practice.
```

The summary is that you should
* Listen on a loopback interface for the authorization response
* Open a browser with the authorization request with PKCE with the redirect uri set with your loopback interface
* Process the authorization response from the loopback interface
* Make the request from your native app to the token endpoint to get the tokens with the code and PKCE code verifier

Some sample code can be found at https://github.com/googlesamples/oauth-apps-for-windows.

**Reply by Mohitkiran** [reply]
Thanks for the reply Justin, this app which we are building works on login screen of the windows and I am not sure if we can open a browser window on login screen due to security reasons. Is there any other way to this? 

But I will try this approach and get back to you if this works.
